### ISSUE TITLE
Mask.to_surface

### ISSUE THREAD

**Issue by illume** [pygame/pygame#1070]
Reading the https://github.com/pygame/pygame/issues/1029 issue again... helpful would be some sort of Mask drawing functions.

Some ideas about a possible API...

```
Mask.to_surface(surface, offset, setcolor=(0,0,0,127), unsetcolor=None)
-> draws setcolor onto the surface where the mask is set to 1.

Mask.to_surface(surface, offset, setcolor=(0,0,0,127), unsetcolor=(255,255,255,127))
-> draws setcolor onto the surface where the mask is set to 1, 
    and unsetcolor where the mask is set to 0.

Mask.to_surface(surface, offset, setsurface=surf, unsetsurface=None)
 -> draws on surface, using values from setsurface where the mask is set to 1.

Mask.to_surface(surface, offset, setsurface=surf, unsetcolor=(255,255,255,127))
 -> draws on surface, using values from setsurface where the mask is set to 1.
    and unsetcolor where the mask is set to 0.

Mask.to_surface(surface=None, offset=None, setsurface=surf)
Mask.to_surface(setsurface=surf)
->creates a new surface because surface=None, based on the setsurface settings.
```

This would allow drawing on an existing surface, as well as making it easy to create a new surface based on the mask.

`to_surface` is used here because it is the opposite of from_surface.

**Comment by illume** [comment]
Another approach to think about is to make a Mask be an acceptable Surface. But this is a __LOT__ more work. The idea being that Surface.blit(mask, ...) and pygame.draw.line(mask, ...) would somehow work.

**Comment by DeluxeYang** [comment]
this would be really helpful

**Comment by charlesej** [comment]
Work items for `mask.to_surface()`.

- [ ] support for `surface`, `setcolor` and `unsetcolor` kwargs
  - [ ] code
  - [ ] tests
  - [ ] documentation

- [ ] support for `setsurface` and `unsetsurface` kwargs
  - [ ] code
  - [ ] tests
  - [ ] documentation

- [ ] support for `offset` kwarg
  - [ ] code
  - [ ] tests
  - [ ] documentation

- [ ] support for creating a new surface, i.e. when `surface=None`
  - [ ] code
  - [ ] tests
  - [ ] documentation

**Comment by illume** [comment]
Thanks for the checklist @charlesej 

I left blending and all that type of stuff out of the API, because it's probably quite a lot of work. The functions here would be like a standard fill/copy of colors.

**Comment by charlesej** [comment]
@illume What are your thoughts with regards to the `offset` parameter? Do you want to use the `offset` to signify where in the mask to start drawing or where on the surface to start drawing?

```python
# Start drawing the mask's position (0, 0) at position (5, 10) on the given surface.
Mask.to_surface(surface, offset=(5, 10), unsetcolor=None)
```

And/or?

```python
# Start drawing the mask's position (5, 10) at position (0, 0) on the created surface.
Mask.to_surface(offset=(5, 10))
```

**Comment by illume** [comment]
Good question. 

For offset, I was looking at the `overlap` method of Mask.
But maybe we should copy the blit `dest`, and `area` arguments instead?

- `area` rectangle represents a smaller portion of the source mask to draw.
- `dest` rectangle is where the drawing on the destination surface should happen. Useful if `surface` is larger than than the mask, or they just want to draw somewhere else on the surface.

**Comment by illume** [comment]
Thanks for the setsurface/unsetsurface support @charlesej . I'm really looking forward to using that.

**Comment by charlesej** [comment]
You are welcome.

There is one thing that I didn't implement with regards to the `setsurface` parameter from your example code.

```
Mask.to_surface(surface=None, setsurface=surf)
Mask.to_surface(setsurface=surf)
->creates a new surface because surface=None, based on the setsurface settings.
```

The `setsurface` settings are currently not used when creating a new surface. The new surface is always created with `flags=SRCALPHA` and `depth=32`. I felt it would be better if the `setsurface` and `unsetsurface` parameters were only used for colors. The user can always pass a `surface=` parameter if they wanted something other than the default surface creation.

Do you still want the `setsurface` settings to be used when creating a new surface? Also if no `setsurface` is provided should the settings from `unsetsurface` be used?

**Comment by illume** [comment]
I agree that people can pass in a surface if they want to specify different flags. Using setsurface for this would be a bit confusing IMHO.

**Comment by illume** [comment]
Just now, reading this code by itself isn't as clear as it could be...

```python
Mask.to_surface(setsurface=surf)
```

`setsurface=surf` sort of looks like we are setting the surface.

Maybe we need to rethink the wording here?

**Comment by charlesej** [comment]
Agreed.

Instead of the `setsurface` / `unsetsurface` keywords how about:

```python
Mask.to_surface(setfrom=surface1)
Mask.to_surface(unsetfrom=surface2)
Mask.to_surface(setfrom=surface1, unsetfrom=surface2)
```

OR

```python
Mask.to_surface(setfromsurf=surface1)
Mask.to_surface(unsetfromsurf=surface2)
Mask.to_surface(setfromsurf=surface1, unsetfromsurf=surface2)
```

**Comment by illume** [comment]
How about these?

```python
Mask.to_surface(offcolor=Color('blue'))
Mask.to_surface(oncolor=Color('red'))

Mask.to_surface(offsurface=surface1)
Mask.to_surface(onsurface=surface1)
```

**Comment by charlesej** [comment]
Using `oncolor` and `oncolors` seems like it might cause some confusion. So for clarity I would favor using `oncolor` / `offcolor` and `onsurface` / `offsurface`.

**Comment by illume** [comment]
Cool.

Thinking about making it blit() api compatible... somehow.

```python
mask.set_color = 'black'
mask.unset_color = 'white'
mask.set_surface = s
mask.unset_surface = s2
mask.blit(surface, (0, 0))
```

Maybe default to set_color == 'black' and everything else off would be ok.
```python
mask.set_color = 'red'
mask.blit(surface, (0, 0))
```

**Comment by charlesej** [comment]
Were you thinking of adding this in addition to the `to_surface()` method or replacing it? My thoughts would be to get the rest of the `to_surface()` functionality (dest and area parameters) working and tested before attempting the `blit` interface.

**Comment by illume** [comment]
I think keeping to_surface is fine and good. Yes, finishing it off first is a better idea.

**Comment by robertpfeiffer** [comment]
So how many steps do I have to take when I want to blit from surf A to surf B, but only the pixels from surf A where the mask is set?

**Comment by illume** [comment]
> mask.to_surface(surfaceB, onsurface=surfaceB)
